<!--[![DOI](https://zenodo.org/badge/829822589.svg)](https://zenodo.org/doi/10.5281/zenodo.13347495)-->
# Overview of the scMnT method
![scMnT preview](https://raw.githubusercontent.com/18parkky/NanoMnT/scMnT/Overview.jpg)
 
# Table of contents
1. [Installation](#installation)<br/>
2. [Commands & Parameters](#commands--parameters)<br/>
3. [Output files](#output-files)<br/>
4. [Using scMnT](#using-scmnt)<br/> 
5. [Tutorial - Identifying MSI status in single-cell resolution](#tutorial---identifying-msi-status-in-single-cell-resolution-using-cancer-cell-data-from-a-study-by-kinker-et-al)<br/>
6. [Citation](#citation)<br/>

# scMnT
scMNT is a specialized extension of NanoMnT, developed for a distinct purpose: quantifying microsatellite instability (MSI) intensity at single-cell resolution. It provides a suite of Python scripts that analyze reads mapped to mononucleotide-repeat microsatellite regions.

## Installation
Before installing, ensure that the following packages are installed:
  
  - Matplotlib >= 3.7.1
  - Numpy >= 1.20.3
  - Pysam >= 0.20.0
  - Pandas >= 2.0.0
  - Scipy >= 1.7.1
  - Seaborn >= 0.13.0

After installing these packages, use the following commands to install scMnT.
*You may consider creating a new Conda environment for scMnT.
```
git clone -b scMnT https://github.com/18parkky/NanoMnT.git
cd NanoMnT
pip install .
```

scMnT requires [Krait](https://github.com/lmdu/krait) output files as inputs, which needs to be installed separately.  
Krait output files for mononucleotide-repeat regions (≥10 bp) are available in scMnT GitHub repository under:`~/NanoMnT/ref/krait/`.
scMnT has been tested in Python version 3.9.7. and should work in later versions.

## Commands & Parameters
scMnT uses three main commands in the following order: `getAlleleTable`, `scMnT-score`, `scMnT-find`, and an important Jupyter notebook script `findMSIcell.ipynb`.

### 1) `getAlleleTable`
The `getAlleleTable` command requires (1) an input BAM and (2) Krait file (to specify STR loci of interest) as inputs, and produces a single TSV file, namely, the allele table. The allele table contains information about every reads that aligned to any of the given STR locus. In scMnT, the default parameters are different from those of NanoMnT due to the different nature of Oxford Nanopore instruments and Illumina instruments. If your sequencing data has been generated with other instruments (e.g., Nanopore or PacBio), change the optional parameters accordingly.

In a typical scRNA-seq data, you will be using the BAM file generated by Cell Ranger (e.g., `possorted_genome_bam.bam`) and the Krait file under `~/NanoMnT/ref/krait/` as the inputs.

```
usage: getAlleleTable.py [-h] -b PATH_BAM -s PATH_STR_TSV -r PATH_REF_GENOME [-m MAPPING_QUALITY] [-f FLANKING] [-rf REALIGNMENT_FLANKING] [-t THREADS] [--sc] [-out DIR_OUT]
```

**Required parameters**:
- `-b`: PATH to input BAM
- `-s`: PATH to Krait output  
- `-r`: PATH to reference genome
   
**Optional parameters**:
- `-m`: Minimum MAPQ when processing reads. Reads with MAPQ below this will be discarded (default: 0)
- `-f`: Length of flanking sequences to be written on allele table (default: 6)
- `-rf`: Length of flanking sequences of STR to use as reference, during re-alignment process (default: 1000)
- `-t`: Number of threads to use for multiprocessing (default: 4)
- `--sc`: Option that tells NanoMnT to look for CB and UMI for each read (default: True)
- `-out`: Directory to write output files (default: current directory)

### 2) `scMnT-score`
The `scMnT-score` command requires (1) an Scanpy object file (h5ad) where a unique identifier for each cell must be availble at `adata.obs['Identifier']` and (2) the allele table generated using `getAlleleTable` as inputs. `scMnT-score` then calculates the MSI score for each cell (saving it to `adata.obs['MSI_score']`) and writes the resulting Scanpy object file to disk.

```
usage: scMnT-score [-h] -a PATH_SCANPY_ADATA -at PATH_STR_ALLELE_TABLE  [-out DIR_OUT]
```
**Required parameters**:
- `-a`: PATH to Scanpy object file (h5ad) with `adata.obs['Identifier']` available.
- `-at`: PATH to allele table
      
**Optional parameters**:
- `-out`: Directory to write output file to (default: current directory)

### 3) `scMnT-find`
The `scMnT-find` command find cell types that are highly likely to be MSI. It requires (1) an Scanpy object file where MSI score (via `scMnT-score`) and the cell type for each cell is labeled and (2) a comma separated list of normal cell types (cells that are surely not MSI, such as monocytes) to use as reference. `scMnT-find` will then produce a TSV file that tells the MSI stauts of each cell type. The cell type for each cell must be available at `adata.obs['CellGroup']`.

```
usage: scMnT-find [-h] -a PATH_SCMNT_SCANPY_ADATA -n NORMAL_CELLTYPES [-m MINIMUM_LOCI] [-out DIR_OUT]
```

**Required parameters**:
- `-a`: PATH to Scanpy object file (h5ad) with both `adata.obs['MSI_score']` and `adata.obs['CellGroup']` available.
- `-n`: a comma separated list of normal cell types. (e.g., monocyte,T,fibroblast)

**Optional parameters**:
- `-m`: Minimum number of MS loci required per cell to be included in the analysis (default: 10). Cells with fewer loci will be excluded from analysis.
- `-out`: Directory to write output files (default: current directory)

## Output files

By default, running each command will each generate one file (+one log file).  

### 1) `getAlleleTable` → allele table
As mentioned above, `getAlleleTable` produces the 'allele table', of which each row contains the following information:
- Read name (`read_name`)
- Locus (`locus`)
- Repeat unit (`repeat_unit`)
- STR allele sequence reported by the read (e.g., AAAAAAAA) (`allele`)
- STR allele (=number of repeats) of the reference genome (`reference_STR_allele`)
- Left flanking sequence of STR (`left_flanking_seq`)
- Right flanking sequence of STR (`right_flanking_seq`)
- Flag of the read (`flag`)
- Cell barcode (`CB`)
- Unique molecular identifier (`UMI`)
- Corrected allele (error-corrected allele) (`corrected_allele`)
- Editing distance between the uncorrected allele and the corrected allele (`editing distance`)
- STR allele reported by the read (`read_STR_allele`)

### 2) `scMnT-score` → Scanpy object file
Running `scMnT-score` creates a Scanpy object file with MSI score labeled under `adata.obs['MSI_score']`.

### 3) `scMnT-find` → TSV file
Running `scMnT-find` creates a TSV file with the following columns: 'CT', 'pval', 'delta', 'n_cells'. 
- CT = Cell type
- pval = the p-value of the null hypothesis — that the MSI score distribution of the given cell type originates from the same distribution as that of the normal cell types — as tested by the Kolmogorov–Smirnov test
- delta = Cliff's delta value
- Number of cells used for the Kolmogorov-Smirnov test

## Using scMnT
Generally, the scMnT workflow is as follows:
1. Cell type annotation (e.g., using Scanpy/Seurat)
2. `getAlleleTable`
3. `scMnT-score` 
4. Either (1) `scMnT-find` or (2) custom analyses (e.g., UMAP inspection of MSI scores and subsequent subclustering).


## Tutorial - Identifying MSI status in single-cell resolution using cancer cell data from a study by [Kinker et al.](https://www.nature.com/articles/s41588-020-00726-6)
We provide a tutorial notebook(`~/NanoMnT/scMnT/workflow.ipynb`) that well-represents the full capability of scMnT.
**We highly recommend the users to go through the tutorial (or even use them for their analyses).**

This tutorial uses the cancer cell line dataset from a study by Kinker et al., guiding the users through:
- Identifying MSI cells
- Estimating MSI intensity
- Visualizing results using UMAP

You are free to reproduce the results using the same data, or you can simply replace the input files with your own to run the workflow.

**Required files**:
- PATH to input BAM (e.g., `possorted_genome_bam.bam`)
- PATH to the Scanpy object file with cell type annotation and MSI score labeled. 

## Citation  
If you use scMnT in your research, please cite our [preprint](https://www.biorxiv.org/content/10.1101/2025.05.09.653227v1).
We will update the DOI when we publish our paper in a peer-reviewed journal.

Like NanoMnT, scMnT is under active development, so feel free to make suggestions!
